File: pt5/1138.alphabet-board-path.java
import java.util.*;

class Solution {
    public String alphabetBoardPath(String target) {
        int row = 0;
        int col = 0;
        StringBuilder moves = new StringBuilder();

        for (char ch : target.toCharArray()) {
            int index = ch - 'a';
            int nextRow = index / 5;
            int nextCol = index % 5;

            while (row > nextRow) {
                moves.append('U');
                row--;
            }
            while (col > nextCol) {
                moves.append('L');
                col--;
            }
            while (col < nextCol) {
                moves.append('R');
                col++;
            }
            while (row < nextRow) {
                moves.append('D');
                row++;
            }
            moves.append('!');
        }

        return moves.toString();
    }
}

File: pt5/1138.alphabet-board-path.py
from __future__ import annotations


class Solution:
    def alphabetBoardPath(self, target: str) -> str:
        row = col = 0
        moves = []

        for char in target:
            index = ord(char) - ord("a")
            next_row, next_col = divmod(index, 5)

            while row > next_row:
                moves.append("U")
                row -= 1
            while col > next_col:
                moves.append("L")
                col -= 1
            while col < next_col:
                moves.append("R")
                col += 1
            while row < next_row:
                moves.append("D")
                row += 1

            moves.append("!")

        return "".join(moves)

